Tests/NumLib/NaturalNodeCoordinates.cpp

```diff
@@ -0,0 +1,363 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2017, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include <gtest/gtest.h>
+
+#include <vector>
+
+#include "MeshLib/ElementCoordinatesMappingLocal.h"
+
+#include "NumLib/Fem/CoordinatesMapping/NaturalNodeCoordinates.h"
+#include "NumLib/Fem/CoordinatesMapping/ShapeMatrices.h"
+#include "NumLib/Fem/FiniteElement/C0IsoparametricElements.h"
+#include "NumLib/Fem/Integration/GaussIntegrationPolicy.h"
+#include "NumLib/Fem/ShapeMatrixPolicy.h"
+
+template <typename ShapeFunction, int GlobalDim>
+bool test(MeshLib::Element const& element)
+{
+    using ShapeMatricesType = ShapeMatrixPolicyType<ShapeFunction, GlobalDim>;
+    using FemType =
+        NumLib::TemplateIsoparametric<ShapeFunction, ShapeMatricesType>;
+
+    // Test if the cast is possible.
+    bool const dynamic_cast_failed =
+        dynamic_cast<typename ShapeFunction::MeshElement const*>(&element) ==
+        nullptr;
+    if (dynamic_cast_failed)
+    {
+        EXPECT_TRUE(dynamic_cast_failed);
+        return false;
+    }
+    FemType fe{
+        static_cast<typename ShapeFunction::MeshElement const&>(element)};
+
+    bool result = true;
+
+    // For each node evaluate the shape functions at natural coordinates and
+    // test if only the corresponding shape function has value 1 and all others
+    // must return 0.
+    int const number_nodes = element.getNumberOfNodes();
+    for (int n = 0; n < number_nodes; ++n)
+    {
+        // Evaluate shape matrices at natural coordinates.
+        typename ShapeMatricesType::ShapeMatrices shape_matrices{
+            ShapeFunction::DIM, GlobalDim, ShapeFunction::NPOINTS};
+        // Compute only N, because for pyramid the detJ becomes zero at the
+        // apex, and we only use N in the following test anyway.
+        fe.template computeShapeFunctions<NumLib::ShapeMatrixType::N>(
+            NumLib::NaturalCoordinates<
+                typename ShapeFunction::MeshElement>::coordinates[n]
+                .data(),
+            shape_matrices, GlobalDim, false /* axial symmetry */);
+
+        auto const& N = shape_matrices.N;
+        for (int p = 0; p < ShapeFunction::NPOINTS; ++p)
+        {
+            if (p == n)
+            {
+                if (N[p] != 1)
+                {
+                    EXPECT_EQ(1, N[p]) << "for n = " << n << ", p = " << p
```

Comment:
Nice feature. I guess that will print an extended message upon failing test.

Reply:
Correct.